### ISSUE TITLE
The behavior of the intersection type of array changed in 2.6.1

### ISSUE THREAD

**Issue by icepeng** [Microsoft/TypeScript#20001]
**TypeScript Version:**  2.6.1

**Code**

```ts
interface A {
  first: string;
  bList?: B[];
}

interface B {
  second: string;
  c?: C;
}

interface C {
  third: string;
}

const Q: A & { bList: (B & { c: C })[] } = {
  first: 'first',
  bList: [
    {
      second: 'second',
      c: {
        third: 'third',
      },
    },
  ],
};

Q.bList.map(item => {
  P(item);
});

function P(value: B & { c: C }) {
  console.log(value);
}

```

**Expected behavior:**
Until 2.5.1 this code did not generate a compilation error.
The type of bList is calculated as follows:
```ts
(undefined & (B & {
  c: C;
})[]) | (B[] & (B & {
  c: C;
})[])
```
And type of item
```ts
B & {
  c: C;
}
```
**Actual behavior:**
With 2.6.x, this code causes an error.
The type of bList is calculated as follows:
```ts
B[] & (B & {
  c: C[];
})[]
```

And type of item
```ts
B
```

The type of item is treated as B and an error occurs. Is this change specified in the changelog?

**Comment by ahejlsberg** [comment]
This is an effect of #18438. We now remove empty intersection types (i.e. intersection types for which no values exist) from union types. In your example, this means that `undefined & (B & { c: C })[]` is removed when constructing the type for `bList`, leaving just `B[] & (B & { c: C })[]`. Then, as we look for a `map` member on that type, we end up picking the candidate from `B[]` and therefore `item` has type `B`. Previously we'd pick a candidate from `(B & { c: C })[]` because we didn't remove the empty intersection.

In general, when the constituent types of intersections have overlapping signatures, the intersection is *order sensitive* because overload resolution processes the signatures from left to right. Indeed, your example works if you change the type of `Q` to list `A` last:

```ts
const Q: { bList: (B & { c: C })[] } & A = { ... };  // Now works 
```

**Comment by icepeng** [comment]
It would be nice to write that the intersection type is order sensitive in the advanced types section of the typescript handbook.
